File: java/src/main/java/com/nedap/retail/api/v1/tester/EventsServerConnectionHandler.java
package com.nedap.retail.api.v1.tester;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

/**
 * Handles a socket connection from the API (HTTP, Content transfer encoding is
 * chunked) and prints the data to System.out
 */
public class EventsServerConnectionHandler implements Runnable {
    /**
     * For reading from the socket
     */
    private BufferedReader in;
    /**
     * For writing to the socket
     */
    private PrintWriter out;
    /**
     * The socket
     */
    private Socket client;

    public EventsServerConnectionHandler(Socket client) {
        this.client = client;
        try {
            this.in = new BufferedReader(new InputStreamReader(client.getInputStream()));
            this.out = new PrintWriter(client.getOutputStream());
        } catch (IOException e) {
            System.out.println(e);
        }
    }
    
    public void run() {
        String message = "";
        boolean readingHeader = true;
        boolean reading = true;
        try {
            while(reading) {
                // wait for input from client
                message = in.readLine();
                if (message==null) {
                    // client is disconnected
                    reading = false;
                } else {
                    if (readingHeader) {
                        // empty line means end of HTTP headers
                        if (message.length()==0) {
                            readingHeader = false;
                        }
                    } else {
                        if (message.equals("0")) {
                            // chunk length 0 means end of request
                            reading = false;
                        } else {
                            System.out.println(message);
                        }
                    }
                }
            }
            // send HTTP reply
            out.println("HTTP/1.1 200 OK");
            out.println("Content-Type: text/html");
            out.println("X-Powered-By: Nedap Retail");
            out.println("Connection: close");
            out.println("Content-Length: 3");
            out.println("");
            out.println("Ok");
            out.close();
        } catch (IOException e) {
            System.out.println(e);
        } finally {
            try {
                client.close();
            } catch (Exception e) {
                
            }
        }
    }
}

File: csharp/EventsServerConnectionHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.IO;
using System.Threading;

namespace StoreIDTopGateAPI
{
    class EventsServerConnectionHandler
    {
        private TcpClient client;
        private EventsServer server;
        private Stream inputStream;
        private StreamWriter outputStream;

        public EventsServerConnectionHandler(TcpClient client, EventsServer server)
        {
            this.client = client;
            this.server = server;
        }

        public void handle()
        {
            inputStream = new BufferedStream(client.GetStream());
            outputStream = new StreamWriter(new BufferedStream(client.GetStream()));
            String message = "";
            Boolean readingHeader = true;
            Boolean reading = true;
            // skip headers
            while (reading && ((message = streamReadLine()) != null))
            {
                if (readingHeader)
                {
                    if (message.Equals(""))
                    {
                        readingHeader = false;
                    }
                }
                else
                {
                    if (message.Equals("0"))
                    {
                        // chunk length 0 means end of request
                        reading = false;
                    }
                    else
                    {
                        Console.WriteLine(message);
                    }
                }
            }

            // send reply
            outputStream.Write("HTTP/1.0 200 OK\n");
            outputStream.Write("Content-Type: text/html\n");
            outputStream.Write("Connection: close\n");
            outputStream.Write("Content-Length: 3\n");
            outputStream.Write("\n");
            outputStream.Write("Ok\n");
            outputStream.Flush();
            outputStream.Close();
            inputStream.Close();
            client.Close();
        }

        private string streamReadLine() {
            int next_char;
            string data = "";
            while (true) {
                next_char = inputStream.ReadByte();
                if (next_char == '\n') { break; }
                if (next_char == '\r') { continue; }
                if (next_char == -1) { Thread.Sleep(1); continue; };
                data += Convert.ToChar(next_char);
            }            
           return data;
       }
    }
}
